example/bwmf/bwmf_task.go

```diff
@@ -168,20 +300,83 @@ func (t *bwmfTask) doDataReady(ctx context.Context, fromID uint64, method string
 
 // These two function carry out actual optimization.
 func (t *bwmfTask) updateDShard(ctx context.Context) {
+	if t.dims.N == -1 {
+		t.dims.N = 0
+		for _, m := range t.peerShards {
+			t.dims.N += len(m.Row)
+		}
+		wh := make([][]float32, t.dims.N)
+		for i, _ := range wh {
+			wh[i] = make([]float32, t.dims.m)
+		}
+		t.dLoss = &KLDivLoss{
+			V:  t.rowShard,
+			WH: wh,
+			m:  t.dims.N,
+			n:  t.dims.m,
+			k:  t.dims.k,
+		}
+	}
+
+	t.dLoss.W = NewBlocksParameter(&t.peerShards)
+	param := NewSingleBlockParameter(t.dShard)
+
+	ok := t.optimizer.Minimize(t.dLoss, t.stopCriteria, param)
+	if !ok {
+		t.logger.Panicf("Failed minimizing over dShard.")
+		// handle re-run
+		// XXX(baigang) just kill the framework and wait for restarting?
+	}
+
+	// XXX(baigang): This should be in updateDShard, i.e in even epochs.
+	t.curIter += 1
+	if t.curIter >= t.numIters {
+		// TODO save result to filesystem
+		// but we can dump it to logger here for now
+		t.logger.Println("tShard: ", *t.tShard)
+		t.logger.Println("dShard: ", *t.dShard)
+		t.logger.Println("Finished. Waiting for the framework to stop the task...")
+		// XXX We do not return here because it's up to the master to stop it and we just report the epoch is done.
+	}
+
 	select {
-	case <-time.After(time.Duration(rand.Intn(3)) * time.Second):
+	case t.updateDone <- &event{ctx: ctx}:
 	case <-ctx.Done():
-		return
 	}
-	t.updateDone <- &event{ctx: ctx}
 }
+
 func (t *bwmfTask) updateTShard(ctx context.Context) {
+	if t.dims.M == -1 {
+		t.dims.M = 0
+		for _, m := range t.peerShards {
+			t.dims.M += len(m.Row)
+		}
+		wh := make([][]float32, t.dims.M)
+		for i, _ := range wh {
+			wh[i] = make([]float32, t.dims.n)
+		}
+		t.tLoss = &KLDivLoss{
+			V:  t.columnShard,
+			WH: wh,
+			m:  t.dims.M,
+			n:  t.dims.n,
+			k:  t.dims.k,
+		}
+	}
+
+	t.tLoss.W = NewBlocksParameter(&t.peerShards)
+	param := NewSingleBlockParameter(t.tShard)
+
+	ok := t.optimizer.Minimize(t.tLoss, t.stopCriteria, param)
```

Comment:
@BaiGang 
I've merged your change on Minimize interface.

Can you fix this too?